File: trunk/ip4addr.cpp
/* ip4addr.cpp - IP address utility
 *
 * Copyright 2011 Patrick Allen
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/* Stores integers in NBO */

#include <arpa/inet.h>
#include <inttypes.h>
#include <cstring>
#include <iostream>

#include "ip4addr.hpp"

namespace IP4Addr
{
        hash_t IP4Addr::makeHash(in_addr_t addr, plen_t plen)
        {
                return ((addr << 31) + plen);
        }

	hash_t IP4Addr::makeHash(in_addr_t addr, in_addr_t snmask)
	{
		/* Since the hash is a network address + prefix length,
		 * if we are given the subnet mask we must first convert it
		 * to a prefix length. */
		plen_t plen = smtopl(snmask);

		/* Call the other hash function */
		return makeHash(addr, plen);
	}
	
	/* valid_snmask_i
	 * Make sure all the bits flipped on are to the 'left' side of the
	 * number.  Otherwise it's not a subnet mask.
	 */
	bool IP4Addr::valid_snmask_i(in_addr_t snmask) const
	{
		char maskchk[] = { 128, 64, 32, 16, 8, 4, 2, 1 };
		char * chkptr = maskchk;
		uint8_t * octet;
		bool hostbits = false;
	
		//snmask = htonl(snmask);
	 	octet = (uint8_t *) &snmask;
	
		/* For every octet, check if the mask is ok */
		for (int i = 0; i < 4; i++, octet++)
		{
			for (chkptr = maskchk; chkptr <= &maskchk[7]; chkptr++)
			{
				if (! hostbits)
				{
					if (*chkptr & *octet)
					{
						continue;
					}
					else
					{
						hostbits++;
					}
				}
				else
				{
					if (*chkptr & *octet)
					{
						/* Bad mask */
						return false;
					}
					else
					{
						continue;
					}
				}
			}
		}

		return true;
	}
	
	/* pltosn
	 *
	 * Prefix length to subnet mask.
	 */
	in_addr_t IP4Addr::pltosm(plen_t plen) const
	{
		uint32_t highbit = 1 << 31;  /* Leftmost bit flipped on */

		/* The following won't work with a prefix length of 0,
		 * so handle that now */
		if (plen == 0)
		{
			return 0;
		}

		/* If the prefix length was greater than the max,
		 * reduce it to 32. */
		if (plen > 32)
		{
			plen = 32;
		}

		/*
	 	 * 1. Zero out MASK, then set MASK's highest order bit equal to one
	 	 * 2. Right shift MASK by one less than the PREFIX LENGTH
	 	 * 3. Subtract 1 from MASK; now all the bits we want are flipped to 0
	 	 * 4. Get the one's complement of MASK (flip all 0s to 1s, vice versa)
	 	 * 5. MASK is now in network byte order; convert it to host order
	 	 */
		return ntohl(~((highbit >> plen - 1) - 1));
	}

	/* smtopl
	 * Subnet mask (as a host order integer) to prefix length conversion.
	 */ 
	plen_t IP4Addr::smtopl(in_addr_t snmask)
	{
		uint8_t * cur = (uint8_t *) &snmask;
		plen_t plen = 0;
	
		/* Convert to network byte order */
		//snmask = htonl(snmask);
	
		/* Check # of bits on in each octet */
		for (int i = 0; i < 4; i++, cur++) {
			if (*cur == 255) {
				/* All bits were on */
				plen += 8;
				continue;
			} else {
				/* Find consecutive leftmost bits on within this byte */
				char bits[] = { 128, 64, 32, 16, 8, 4, 2, 1 };
				char * bptr = &bits[0];
				char * startbit = &bits[0];
	
				while (*cur & *bptr)
					bptr++;
				plen += (bptr - startbit);
				break;
			}
		}
	
		return plen;
	}
	
	/* Get functions */
	
	in_addr_t IP4Addr::getAddr(char * buf)
	{
		if (! isValid())
		{
			return 0;
		}

		if (buf != NULL)
		{
			strncpy(buf, m_addr_s, INET_ADDRSTRLEN);
		}

		return m_addr_i;
	}

	in_addr_t IP4Addr::getSnmask(char * buf)
	{
		if (! isValid())
		{
			return NULL;
		}

		if (buf != NULL)
		{
			strncpy(buf, m_snmask_s, INET_ADDRSTRLEN);
		}

		return m_snmask_i;
	}

	hash_t IP4Addr::getHash()
	{
		if (! isValid())
		{
			return 0;
		}

		return m_hash;
	}

	plen_t IP4Addr::getPlen()
	{
		if (! isValid())
		{
			return 0;
		}

		return m_plen;
	}

	bool IP4Addr::isValid()
	{
		return (addrValid && maskValid);
	}

	/* Gives both the string and integer (NBO) representation of a
	 * broadcast address. String goes in the buf if it is not NULL.
	 * Integer is returned normally.
	 */
	in_addr_t IP4Addr::getBroadcast(char * buf)
	{
		in_addr_t bc;

		if (! isValid())
		{
			return 0;
		}

		bc = m_addr_i | getHostmask(NULL);

		if (buf != NULL)
		{
			int hboBc = ntohl(bc);
			inet_ntop(AF_INET, &hboBc, buf, INET_ADDRSTRLEN);
		}

		return bc;
	}

	in_addr_t IP4Addr::getHostmask(char * buf)
	{
		in_addr_t hmask;

		if (! isValid())
		{
			return 0;
		}

		hmask = ~ getSnmask(0);

		if (buf != NULL)
		{
			int hboHmask = ntohl(hmask);
			inet_ntop(AF_INET, &hboHmask, buf, INET_ADDRSTRLEN);
		}

		return hmask;
	}

	in_addr_t IP4Addr::getNetwork(char * buf)
	{
		in_addr_t net;

		if (! isValid())
		{
			return 0;
		}

		net = m_snmask_i & m_addr_i;

		if (buf != NULL)
		{
			int hboNet = ntohl(net);
			inet_ntop(AF_INET, &hboNet, buf, INET_ADDRSTRLEN);
		}

		return net;
	}

	bool IP4Addr::setAddr_s(char * addr_s)
	{
		if (! addr_s)
		{
			return setAddrFail();
		}

		if (! inet_pton(AF_INET, addr_s, &m_addr_i))
		{
			return setAddrFail();
		}

		/* Convert address to NBO */
		m_addr_i = htonl(m_addr_i);

		strncpy(m_addr_s, addr_s, INET_ADDRSTRLEN);
		return setAddrSuccess();
	}

	bool IP4Addr::setAddr_i(in_addr_t addr_i)
	{
		if (! (inet_ntop(AF_INET, &addr_i, m_addr_s, INET_ADDRSTRLEN)))
		{
			return setAddrFail();
		}

		m_addr_i = htonl(addr_i);
		return setAddrSuccess();
	}

	bool IP4Addr::setSnmask_s(char * snmask_s)
	{
		if (! inet_pton(AF_INET, snmask_s, &m_snmask_i))
		{
			return setMaskFail();
		}

		strncpy(m_snmask_s, snmask_s, INET_ADDRSTRLEN);
		m_plen = smtopl(m_snmask_i);
		return setMaskSuccess();
	}

	bool IP4Addr::setSnmask_i(in_addr_t snmask_i)
	{
		if (! (inet_ntop(AF_INET, &snmask_i, m_snmask_s,
		       INET_ADDRSTRLEN)))
		{
			return setMaskFail();
		}

		m_plen = smtopl(snmask_i);
		m_snmask_i = htonl(snmask_i);
		return setMaskSuccess();
	}

	bool IP4Addr::setPlen(int plen)
	{
		if (plen < 0 || plen > 32)
		{
			return setMaskFail();
		}

		setSnmask_i(pltosm(plen));

		return setMaskSuccess();
	}

	bool IP4Addr::setAddrSuccess()
	{
		if (maskValid == true)
		{
			updateNetInfo();
		}

		return addrValid = true;
	}

	bool IP4Addr::setAddrFail()
	{
		return addrValid = false;
	}

	bool IP4Addr::setMaskSuccess()
	{
		if (addrValid == true)
		{
			updateNetInfo();
		}

		return maskValid = true;
	}

	bool IP4Addr::setMaskFail()
	{
		return maskValid = false;
	}

	/* Compute any members that depend on the mask and address */
	void IP4Addr::updateNetInfo()
	{
		makeHash(m_addr_i, m_snmask_i);
		return;
	}

	/* CTORs -- as above */
	IP4Addr::IP4Addr(void)
	{
		setAddrFail();
		setMaskFail();
	}

	IP4Addr::IP4Addr(char * addr_s, char * snmask_s)
	{
		setAddr_s(addr_s);
		setSnmask_s(snmask_s);
	}

	IP4Addr::IP4Addr(in_addr_t addr_i, in_addr_t snmask_i)
	{
		setAddr_i(addr_i);
		setSnmask_i(snmask_i);
	}

	IP4Addr::IP4Addr(char * addr_s, plen_t plen)
	{
		setAddr_s(addr_s);
		setPlen(plen);
	}

	IP4Addr::IP4Addr(in_addr_t addr_i, plen_t plen)
	{
		setAddr_i(addr_i);
		setPlen(plen);
	}

	/* Other */

	std::ostream & operator<<(std::ostream & os, IP4Addr & ip)
	{
		char addr[INET_ADDRSTRLEN];

		if (ip.isValid() == false)
		{
			os << "IP and/or mask is not valid." << std::endl;
			return os;
		}

		ip.getAddr(addr);

		os << "IP Address:    " << addr << " (" << ip.getAddr(NULL)
		   << ")" << std::endl;

		os << "Prefix length: " << (int) ip.getPlen() << std::endl;

		ip.getSnmask(addr);

		os << "Subnet Mask:   " << addr << " (" << ip.getSnmask(NULL)
		   << ")" << std::endl;

		ip.getBroadcast(addr);

		os << "Broadcast:     " << addr << " (" << ip.getBroadcast(NULL)
		   << ")" << std::endl;

		ip.getNetwork(addr);

		os << "Network:       " << addr << " (" << ip.getNetwork(NULL)
		   << ")" << std::endl;

		ip.getHostmask(addr);

		os << "Hostmask:      " << addr << " (" << ip.getHostmask(NULL)
		   << ")" << std::endl;

		return os;
	}

	/* with_snmask
	 * Return the network ANDed with a new network mask.
	 */
	in_addr_t IP4Addr::withSnmask(in_addr_t snmask_i) const
	{
		if (! valid_snmask_i(snmask_i)) {
			return 0;
		}
	
		return m_addr_i & snmask_i;
	}
	
	/* with_snmask
	 * Return the network ANDed with a new network mask.
	 */
	in_addr_t IP4Addr::withSnmask(plen_t plen) const
	{
		if (plen > 32) {
			plen = 32;
		}
	
		in_addr_t snmask = pltosm(plen);
	
		return withSnmask(snmask);
	}
};

File: trunk/python/ip4addr.py
# IP version 4 address class
# Copyright (C) 2011  Patrick F. Allen
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
# USA.

# This uses socket.inet_pton, socket.ntohl, etc, which may or may not be available in
# Windows cpython.
#
# Addresses and mask are set either using a string in dotted decimal format,
# a host order integer, or (for masks) an integer prefix length.
#
# Once an address and mask is set, use the isValid() method to check if
# everything is valid. If it returns True, you can use the get*() methods
# to retrieve other info like the host mask, prefix length, broadcast and
# network addresses, etc., in both string and network order integer form.
#
# Each get*() method returns a tuple. The first item is
# a string in dotted decimal format, and the second is a *network* order
# integer. (Don't be thrown off that you set it in host order and retrieve
# in network order. This is the most useful way to do it for network
# programming. If you want to use values from one IP4Addr to make a new
# IP4Addr, use the string item in the tuple.)

import struct
import socket

class IP4Addr(object):
    def getAddr(self):
        if (self.isValid() == False):
            return None, None

        return self._addr_s, self._addr_i

    def getMask(self, buf = None):
        if (self.isValid() == False):
            return None, None

        return self._mask_s, self._mask_i

    def getBroadcast(self):
        if (self.isValid() == False):
            return None, None

        bc_i = self._addr_i | self.getHostmask()[1]
        bc_s = socket.ntohl(bc_i)
        bc_s = socket.inet_ntop(socket.AF_INET, struct.pack("L", bc_s)[0:4])

        return bc_s, bc_i

    def getHostmask(self):
        if (self.isValid() == False):
            return None, None

        # AND with 0xFFFFFFFF in case of different length integers
        hmask_i = ~ self.getMask()[1] & 0xFFFFFFFF
        hmask_s = socket.ntohl(hmask_i)
        hmask_s = socket.inet_ntop(socket.AF_INET, struct.pack("L", hmask_s)[0:4])

        return hmask_s, hmask_i

    def getNetwork(self):
        if (self.isValid() == False):
            return None, None

        net_i = self._mask_i & self._addr_i
        net_s = socket.ntohl(net_i)
        net_s = socket.inet_ntop(socket.AF_INET, struct.pack("L", net_s)[0:4])

        return net_s, net_i

    def getPlen(self):
        if (self.isValid() == False):
            return None

        return self._plen

    def setPlen(self, plen):
        if (not IP4Addr.isValidPlen(plen)):
            return self._setMaskFail()

        self._plen = plen 
        mask_hostorder = socket.htonl(IP4Addr.pltosm(self._plen))
        self._mask_s = socket.inet_ntop(socket.AF_INET, struct.pack("L", mask_hostorder)[0:4])
        self._mask_i = socket.htonl(mask_hostorder)
        return self._setMaskSuccess()

    @staticmethod
    def add_octets(addrstr):
        addrbytes = bytearray(addrstr)
        return (addrbytes[0] << 24) + \
               (addrbytes[1] << 16) + \
               (addrbytes[2] << 8)  + \
               (addrbytes[3])

    # Mask may be either a subnet mask (in dotted decimal form as a string,
    # or a struct.packed 32 bit address)[0:4], or a prefix length.
    #
    # We'll determine if it's a mask or prefix length by attempting to use
    # it as a subnet mask first; if that fails, we'll try using it as a
    # prefix length.
    #
    # So, the following are all valid and equivalent:
    #    IP4Addr("192.168.0.1", 24)
    #    IP4Addr("192.168.0.1", "255.255.255.0")
    #    IP4Addr(16820416, 16777215)   # Host order integers
    #
    # If no mask is given, a 32 bit mask is assumed
    #    IP4Addr("192.168.0.1") -> Assumes 32 bit mask
    #    IP4Addr(16820416) ->  # Host order integer with assumed 32 bit mask
    # 
    def __init__(self, addr, mask = 32):
        self.setAddr(addr)
        self.setMask(mask)

    def setMask(self, mask):
        if (isinstance(mask, int) or isinstance(mask, long)):
            if (mask < 0):
                return self._setMaskFail()

            if ((IP4Addr.isValidPlen(mask)) == True):
                return self.setPlen(mask)
            else:
                # It's not a prefix length, try using as a subnet mask
                try:
                    self._mask_s = socket.inet_ntop(socket.AF_INET, struct.pack("L", mask)[0:4])
                    self._mask_i = socket.htonl(mask & 0xFFFFFFFF)
                    if (IP4Addr.isValidMask(self._mask_i) == False):
                        return self._setMaskFail()

                    self._plen = IP4Addr.smtopl(self._mask_i)
                    return self._setMaskSuccess()
                except socket.error:
                    return self._setMaskFail()
        elif (isinstance(mask, str)):
            try:
                maskbytes = socket.inet_pton(socket.AF_INET, mask)
                self._mask_i = IP4Addr.add_octets(maskbytes)
                self._mask_s = mask
                self._plen = IP4Addr.smtopl(self._mask_i)
                return self._setMaskSuccess()
            except socket.error:
                return self._setMaskFail()
        else:
            # Bad type, should throw an exception
            return self._setMaskFail()

        # Not reached in setMask()
        assert(False)

    # Subnet mask to prefix length
    @staticmethod
    def smtopl(mask):
        plen = 0

        if ((isinstance(mask, int) != True) and
            (isinstance(mask, long) != True)):
            return None

        # Load the value of each byte into a list of bytes
        maskbytes = []
        maskbytes.append((mask & 0xFF000000) >> 24)
        maskbytes.append((mask & 0x00FF0000) >> 16)
        maskbytes.append((mask & 0x0000FF00) >> 8 )
        maskbytes.append(mask & 0x000000FF)

        for maskbyte in maskbytes:
            if (maskbyte == 255):
                # All bits were on
                plen += 8
            else:
                # Find which bits are on in this octet
                bits = [ 128, 64, 32, 16, 8, 4, 2, 1 ]

                for bit in bits:
                    if (bit & maskbyte == 0):
                        break
                    plen += 1

        return plen

    # Prefix length to subnet mask in network order
    @staticmethod
    def pltosm(plen):
        if (not IP4Addr.isValidPlen(plen)):
            return None

        assert(plen <= 32 and plen >= 0)

        if (plen == 0):
            return 0

        highbit = 1 << 31    # Highest bit flipped on

        # 1.  Begin with a mask of the highest bit flipped on
        # 2.  Right shift mask by one less than the prefix length
        # 3.  Subtract 1 from mask; now all bits we want are 0
        # 4.  Get the one's complement of mask
        # 4a. AND with 0xFFFFFFFF to counter python's type system
        #     (will appear as a negative number otherwise)
        return (~((highbit >> plen - 1) - 1) & 0xFFFFFFFF)

    @staticmethod
    def isValidMask(mask):
        if (isinstance(mask, int)):
            mask = long(mask)

        if (not isinstance(mask, long)):
            return False

        if (mask < 0 or mask > socket.inet_pton(socket.AF_INET, "255.255.255.255")):
            return False

        bits = [ 128, 64, 32, 16, 8, 4, 2, 1 ]
        hostbits = False

        # For every octet, check if mask is ok
        for maskbyte in [ ord(x) for x in struct.pack(">I", mask)[0:4] ]:
            for bit in bits:
                if (hostbits == False):
                    if (maskbyte & bit):
                        continue
                    else:
                        hostbits = True
                else:
                    if (maskbyte & bit):
                        # Bad mask
                        return False
                    else:
                        continue

        return True

    @staticmethod
    def isValidPlen(plen):
        if (not isinstance(plen, int)):
            return False

        if (plen > 32 or plen < 0):
            return False

        return True

    def setAddr(self, addr):
        if (isinstance(addr, int)):
            try:
                self._addr_s = socket.inet_ntop(socket.AF_INET, struct.pack("L", addr)[0:4])
                self._addr_i = socket.htonl(addr)
                return self._setAddrSuccess()
            except socket.error:
                return self._setAddrFail()
        elif (isinstance(addr, str)):
            try:
                self._addr_s = addr
                self._addr_i = socket.inet_pton(socket.AF_INET, self._addr_s)
                self._addr_i = IP4Addr.add_octets(self._addr_i)

                return self._setAddrSuccess()
            except socket.error:
                return self._setAddrFail()
        else:
            # Bad type, should throw an exception
            raise TypeError()

    def _setAddrSuccess(self):
        self._addr_valid = True
        return True

    def _setAddrFail(self):
        self._addr_valid = False
        return False

    def _setMaskSuccess(self):
        self._mask_valid = True
        return True

    def _setMaskFail(self):
        self._mask_valid = False
        return False

    def isValid(self):
        return self._addr_valid and self._mask_valid

    def __str__(self):
        if (self.isValid() == False):
            return "Address not valid."
        return ("%s/%d" % (self.getNetwork()[0], self.getPlen()))

    def printAll(self):
        ret = self.isValid()

        print "Valid:", self.isValid()

        if (ret == True):
            print "Addr:", self.getAddr()
            print "Mask:", self.getMask()
            print "Bcast:", self.getBroadcast()
            print "Hmask:", self.getHostmask()
            print "Net:", self.getNetwork()
            print "Plen:", self.getPlen()
